### ISSUE TITLE
Return explicit ui5 dependencies config

### ISSUE THREAD

**Issue by ThePlenkov** [SAP/ui5-tooling#851]
Hi!

I've noticed that since version 3.0 dependencies now are resolved [via package.json](https://sap.github.io/ui5-tooling/stable/updates/migrate-v3/#changes-to-dependency-configuration).

I just wanted to give you the feedback that this kind of dependency collection will not work in case of monorepos.

If we use monorepo - we maintain all dependencies only on the root level. While in workspaces we use very minimalistic package.json without any dependencies ( they got [automatically ](https://nx.dev/core-features/automate-updating-dependencies)collected during build ) 

Moreover in TypeScript based scenario mapping can be resolved via [tsconfig.json paths](https://www.typescriptlang.org/tsconfig#paths). 

That's why the idea to have ui5 middleware dependencies were not bad actually and not that useless.  It used probably native [require.resolve()](https://nodejs.org/api/modules.html#requireresolverequest-options) and supported all mentioned above techniques.

**Comment by ThePlenkov** [comment]
Additionally - it seems that [new way](https://sap.github.io/ui5-tooling/v3/pages/Workspace/#dependency-management) of dependency resolution using ui5 workspaces also doesn't support  module paths.

For example:
```
Error Message:
Failed to resolve workspace dependency resolution path @sap/ux-ui5-tooling to /workspaces/fiori-tools/test/@sap/ux-ui5-tooling: ENOENT: no such file or directory, open '/workspaces/fiori-tools/test/@sap/ux-ui5-tooling/package.json'
```

I'd suggest that we use relative path only if dot is used. Like `./@sap/ux-ui5-tooling`

**Comment by ThePlenkov** [comment]
@RandomByte correct. I've been using ui5.dependencies in package.json.  I haven't tried yet workspace - because I don't understand what is it for. Is it also for middleware resolution?

**Comment by ThePlenkov** [comment]
So far my workaround is to use * versions in a package.json like here:
https://github.com/theplenkov-npm/fiori-tools/blob/main/test/package.json

while real paths defined here:
https://github.com/theplenkov-npm/fiori-tools/blob/1aa76403b4dd1d7bda89a94e0089505fd8cc6079/tsconfig.base.json#L17

and I run ui5 cli in typescript mode using this trick:
```
#!/usr/bin/env -S node --loader ts-node/esm
import '@ui5/cli/bin/ui5.cjs';
```

so it allows me to run node with ts loader and automatically *.js files got resolved to *.ts

But yes - I'd like still prefer to control this manually. Because middleware load may impact some runtime.

It's not convenient to enable/disable extensions by modifying package.json

**Comment by ThePlenkov** [comment]
Npm doesn't install any dependencies from the registry if it's a workspace package. It just creates symlink. Basically it's similar to file: dependency, just no need to care about path.
And what about middleware resolution- it is controlled by typescript of course.

**Comment by ThePlenkov** [comment]
If we check how different frameworks allow extensibility:
- [Typescript](https://www.typescriptlang.org/tsconfig#plugins)
- [TS-node](https://github.com/TypeStrong/ts-node#paths-and-baseurl)
- [NX](https://nx.dev/plugin-features/use-task-executors#executor-definitions)

almost everywhere plugins are maintained explicitly via config. Having dependency should not mean that this module should be imported automatically.

It's nice also that you have mentioned performance. With a straightforward declaration of middleware list it is not an issue at all, but with automatic scan indeed may appear an issue - especially when we talk about projects with long dependency list

I highly recommend to think of monorepo approach when you will control all the dependencies on a root project level, and package dependencies will be determined during build operation. You will see that in this scenario analyzing dependencies  may not work

**Comment by ThePlenkov** [comment]
No - it behaves same, but I don't like it anyway. Dependency instruction should not be used as a use instruction to my opinion. We need to enable it somewhere explicitly.  Can't we introduce something like .ui5rc file with ability to extend and include plugins?

Similar to what eslintrc is doing or tsconfig

**Comment by RandomByte** [comment]
Thank you very much for your feedback. I'm afraid I don't share your opinion on this matter, but I'll bring it up for discussion in our team anyways. This might take a while though due to the current holiday season.

We actually already introduced a `.ui5rc` file as part of https://github.com/SAP/ui5-tooling/blob/main/rfcs/0013-configuration.md. However, this only contains general (project independent) UI5 Tooling configuration.

I think if we were to add an explicit dependency declaration that overrules the `package.json` dependency, we should do so in the `ui5.yaml`.

**Comment by dfenerski** [comment]
The [`*` workaround](https://github.com/theplenkov-npm/fiori-tools/blob/main/test/package.json) is nice.

@RandomByte how viable is it for the UI5 CLI to "traverse the directories upwards (up to a max depth?) to try to find a 
npm-workspace-`package.json`" and to try inferring dependencies from there?

Maybe this is separate from the original issue. However native npm workspace support is a nice feature to have IMO.

**Comment by RandomByte** [comment]
@dfenerski not sure what kind of setup you have there. With npm workspaces, my understanding is that each workspace package should still declare all required dependencies in its respective package.json (otherwise you might as well remove it and use the root package.json for everything). This kind of setup should already work fine with UI5 CLI. Kindly clarify what I'm missing.

**Comment by dfenerski** [comment]
@RandomByte 

> each workspace package should still declare all required dependencies in its respective package.json

you're right, re-declaration is a good idea.

In my case a workspace package managed to find a monorepo dependency without explicitly declaring it as such.
I wondered why the UI5 CLI does not find it and didn't actually pause to think whether it _should_ 😁
